File: app/src/main/java/com/example/vishistvarugeese/memory/FaceRecognitionActivity.java
package com.example.vishistvarugeese.memory;

import android.Manifest;
import android.app.ProgressDialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.os.Environment;
import android.provider.MediaStore;
import android.support.annotation.NonNull;
import android.support.v4.content.FileProvider;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.View;
import android.widget.EditText;
import android.widget.Toast;

import java.io.File;
import java.io.IOException;

import okhttp3.MediaType;
import okhttp3.MultipartBody;
import okhttp3.OkHttpClient;
import okhttp3.RequestBody;
import okhttp3.ResponseBody;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;

public class FaceRecognitionActivity extends AppCompatActivity {

    static final int REQUEST_TAKE_PHOTO = 1;
    private static final int REQUEST_WRITE_PERMISSION = 786;
    String mCurrentPhotoPath;
    String imageName;

    String name;

    ProgressDialog mProgressDialog;

    Service service;

    private static final int MY_CAMERA_REQUEST_CODE = 100;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_face_recognition);

        overridePendingTransition(R.anim.right_to_left_slide_in, R.anim.right_to_left_slide_out);

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {

            if (checkSelfPermission(Manifest.permission.CAMERA)
                    != PackageManager.PERMISSION_GRANTED) {
                requestPermissions(new String[]{Manifest.permission.CAMERA,
                                Manifest.permission.WRITE_EXTERNAL_STORAGE},
                        MY_CAMERA_REQUEST_CODE);
            }

        }
    }

    public void onRegisterFaceClick(View view) {
        final AlertDialog.Builder alert = new AlertDialog.Builder(this);
        final EditText edittext = new EditText(getApplicationContext());
        alert.setTitle("Enter person's name: ");

        alert.setView(edittext);

        alert.setPositiveButton("TAKE PHOTO", new DialogInterface.OnClickListener() {
            public void onClick(DialogInterface dialog, int whichButton) {
                name = edittext.getText().toString();
                Log.d("mCurrent",name);
                takePhoto();
            }
        });

        alert.setNegativeButton("CANCEL", new DialogInterface.OnClickListener() {
            public void onClick(DialogInterface dialog, int whichButton) {

            }
        });

        alert.show();
    }

    public void onRecognizeFaceClick(View view) {
        name = "1picture";
        takePhoto();
    }

    public void takePhoto() {
        Intent takePictureIntent = new Intent(MediaStore.ACTION_IMAGE_CAPTURE);
        // Ensure that there's a camera activity to handle the intent
        if (takePictureIntent.resolveActivity(getPackageManager()) != null) {
            // Create the File where the photo should go
            File photoFile = null;
            try {
                photoFile = createImageFile();
            } catch (IOException ex) {
                // Error occurred while creating the File
                Log.d("FILE CREATION ERROR", ex + "");
            }
            // Continue only if the File was successfully created
            if (photoFile != null) {
                Uri photoURI = FileProvider.getUriForFile(this,
                        "com.example.vishistvarugeese.memory.fileprovider",
                        photoFile);
                takePictureIntent.putExtra(MediaStore.EXTRA_OUTPUT, photoURI);
                startActivityForResult(takePictureIntent, 1);
            }
        }
    }

    private File createImageFile() throws IOException {
        // Create an image file name
        File storageDir = getExternalFilesDir(Environment.DIRECTORY_PICTURES);
        File image = File.createTempFile(
                name,  /* prefix */
                ".jpg",         /* suffix */
                storageDir      /* directory */
        );
        imageName = image.getName();
        imageName = imageName.substring(0,imageName.length()-4);

        Log.d("mCurrent", imageName);
        // Save a file: path for use with ACTION_VIEW intents
        mCurrentPhotoPath = image.getAbsolutePath();
        Log.d("mCurrentPhoto", mCurrentPhotoPath);
        return image;
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == REQUEST_TAKE_PHOTO && resultCode == RESULT_OK) {
            sendImageToServerDatabase();
        }
    }

    private void sendImageToServerDatabase(){

        String FILE_DIR = "/storage/emulated/0/Android/data/com.example.vishistvarugeese.memory/files/Pictures/";
        Log.d("mCurrentPhoto1", FILE_DIR + imageName);
        File f = new File(FILE_DIR + imageName + ".jpg");

        HttpLoggingInterceptor interceptor = new HttpLoggingInterceptor();
        interceptor.setLevel(HttpLoggingInterceptor.Level.BODY);
        OkHttpClient client = new OkHttpClient.Builder().addInterceptor(interceptor).build();

// Change base URL to your upload server URL.
        service = new Retrofit.Builder().baseUrl("http://192.168.43.110:5000/").client(client).build().create(Service.class);


        RequestBody reqFile = RequestBody.create(MediaType.parse("image/*"), f);
        MultipartBody.Part body = MultipartBody.Part.createFormData("test", f.getName(), reqFile);
        RequestBody name = RequestBody.create(MediaType.parse("text/plain"), "upload_test");

        retrofit2.Call<okhttp3.ResponseBody> req = service.postImage(body, name);
        mProgressDialog = new ProgressDialog(FaceRecognitionActivity.this);
        mProgressDialog.setTitle("Loading...");
        mProgressDialog.setMessage("Please wait");
        mProgressDialog.show();
        req.enqueue(new Callback<ResponseBody>() {
            @Override
            public void onResponse(Call<ResponseBody> call, Response<ResponseBody> response) {
                try {
                    mProgressDialog.dismiss();
                    String responseStr = response.body().string();
                    responseDialog(responseStr);

                } catch (IOException e) {
                    e.printStackTrace();
                }
                Log.d("Response Flask", response.message() + "");
            }

            @Override
            public void onFailure(Call<ResponseBody> call, Throwable t) {
                t.printStackTrace();
                mProgressDialog.dismiss();
                Toast.makeText(getApplicationContext(), t + "", Toast.LENGTH_LONG).show();
            }
        });
    }

    public void responseDialog(String response){
        final AlertDialog.Builder alert = new AlertDialog.Builder(this);
        alert.setTitle(response);

        alert.setNegativeButton("OK", new DialogInterface.OnClickListener() {
            public void onClick(DialogInterface dialog, int whichButton) {

            }
        });

        alert.show();
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == MY_CAMERA_REQUEST_CODE) {
            if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                Toast.makeText(this, "Camera Permission Granted", Toast.LENGTH_LONG).show();
            } else {
                Toast.makeText(this, "Camera Permission Denied", Toast.LENGTH_LONG).show();
            }
        }
    }


    public void onBackPressed(View view) {
        startActivity(new Intent(this, DashboardActivity.class));
        finish();
        overridePendingTransition(R.anim.left_to_right_slide_in, R.anim.left_to_right_slide_out);
    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
        startActivity(new Intent(this, DashboardActivity.class));
        finish();
        overridePendingTransition(R.anim.left_to_right_slide_in, R.anim.left_to_right_slide_out);
    }
}

File: app/src/main/java/com/example/vishistvarugeese/memory/faceRecognition/FaceRecognitionActivity.kt
package com.example.vishistvarugeese.memory.faceRecognition

import android.Manifest
import android.app.ProgressDialog
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import android.util.Log
import android.view.View
import android.widget.EditText
import android.widget.Toast
import com.example.vishistvarugeese.memory.R

class FaceRecognitionActivity : AppCompatActivity() {
    var mCurrentPhotoPath: String? = null
    var imageName: String? = null
    var name: String? = null
    var mProgressDialog: ProgressDialog? = null
    var service: Service? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_face_recognition)
        overridePendingTransition(R.anim.right_to_left_slide_in, R.anim.right_to_left_slide_out)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            if (checkSelfPermission(Manifest.permission.CAMERA)
                    != PackageManager.PERMISSION_GRANTED) {
                requestPermissions(arrayOf(Manifest.permission.CAMERA,
                        Manifest.permission.WRITE_EXTERNAL_STORAGE),
                        MY_CAMERA_REQUEST_CODE)
            }
        }
    }

    fun onRegisterFaceClick(view: View?) {
        val alert = AlertDialog.Builder(this)
        val edittext = EditText(applicationContext)
        alert.setTitle("Enter person's name: ")
        alert.setView(edittext)
        alert.setPositiveButton("TAKE PHOTO") { _, _ ->
            name = edittext.text.toString()
            Log.d("mCurrent", name)
//            takePhoto()
        }
        alert.setNegativeButton("CANCEL") { _, _ -> }
        alert.show()
    }

    fun onRecognizeFaceClick(view: View?) {

    }

//    private fun takePhoto() {
//        val takePictureIntent = Intent(MediaStore.ACTION_IMAGE_CAPTURE)
//        // Ensure that there's a camera activity to handle the intent
//        if (takePictureIntent.resolveActivity(packageManager) != null) {
//            // Create the File where the photo should go
//            var photoFile: File? = null
//            try {
//                photoFile = createImageFile()
//            } catch (ex: IOException) {
//                // Error occurred while creating the File
//                Log.d("FILE CREATION ERROR", ex.toString() + "")
//            }
//            // Continue only if the File was successfully created
//            if (photoFile != null) {
//                val photoURI = FileProvider.getUriForFile(this,
//                        "com.example.vishistvarugeese.memory.fileprovider",
//                        photoFile)
//                takePictureIntent.putExtra(MediaStore.EXTRA_OUTPUT, photoURI)
//                startActivityForResult(takePictureIntent, 1)
//            }
//        }
//    }

//    @Throws(IOException::class)
//    private fun createImageFile(): File {
//        // Create an image file name
//        val storageDir = getExternalFilesDir(Environment.DIRECTORY_PICTURES)
//        val image = File.createTempFile(
//                name,  /* prefix */
//                ".jpg",  /* suffix */
//                storageDir /* directory */
//        )
//        imageName = image.name
//        imageName = imageName.substring(0, imageName.length - 4)
//        Log.d("mCurrent", imageName)
//        // Save a file: path for use with ACTION_VIEW intents
//        mCurrentPhotoPath = image.absolutePath
//        Log.d("mCurrentPhoto", mCurrentPhotoPath)
//        return image
//    }
//
//    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent) {
//        if (requestCode == REQUEST_TAKE_PHOTO && resultCode == Activity.RESULT_OK) {
//            sendImageToServerDatabase()
//        }
//    }
//
//    private fun sendImageToServerDatabase() {
//        val FILE_DIR = "/storage/emulated/0/Android/data/com.example.vishistvarugeese.memory/files/Pictures/"
//        val f = File("$FILE_DIR$imageName.jpg")
//        val interceptor = HttpLoggingInterceptor()
//        interceptor.level = HttpLoggingInterceptor.Level.BODY
//        val client = OkHttpClient.Builder().addInterceptor(interceptor).build()
//
//// Change base URL to your upload server URL.
//        service = Retrofit.Builder().baseUrl("http://192.168.43.110:5000/")
//                .client(client).build()
//                .create(Service::class.java)
//        val reqFile = RequestBody.create(MediaType.parse("image/*"), f)
//        val body = MultipartBody.Part.createFormData("test", f.name, reqFile)
//        val name = RequestBody.create(MediaType.parse("text/plain"), "upload_test")
//        val req = service?.postImage(body, name)
//        mProgressDialog = ProgressDialog(this@FaceRecognitionActivity)
//        mProgressDialog?.setTitle("Loading...")
//        mProgressDialog?.setMessage("Please wait")
//        mProgressDialog?.show()
//        req?.enqueue(object : Callback<ResponseBody> {
//            override fun onResponse(call: Call<ResponseBody>, response: Response<ResponseBody>) {
//                try {
//                    mProgressDialog!!.dismiss()
//                    val responseStr = response.body()!!.string()
//                    responseDialog(responseStr)
//                } catch (e: IOException) {
//                    e.printStackTrace()
//                }
//                Log.d("Response Flask", response.message() + "")
//            }
//
//            override fun onFailure(call: Call<ResponseBody>, t: Throwable) {
//                t.printStackTrace()
//                mProgressDialog!!.dismiss()
//                Toast.makeText(applicationContext, t.toString() + "", Toast.LENGTH_LONG).show()
//            }
//        })
//    }

    fun responseDialog(response: String?) {
        val alert = AlertDialog.Builder(this)
        alert.setTitle(response)
        alert.setNegativeButton("OK") { _, _ -> }
        alert.show()
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == MY_CAMERA_REQUEST_CODE) {
            if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                Toast.makeText(this, "Camera Permission Granted", Toast.LENGTH_LONG).show()
            } else {
                Toast.makeText(this, "Camera Permission Denied", Toast.LENGTH_LONG).show()
            }
        }
    }

    fun onBackPressed(view: View?) {
        finish()
        overridePendingTransition(R.anim.left_to_right_slide_in, R.anim.left_to_right_slide_out)
    }

    override fun onBackPressed() {
        super.onBackPressed()
        finish()
        overridePendingTransition(R.anim.left_to_right_slide_in, R.anim.left_to_right_slide_out)
    }

    companion object {
        const val REQUEST_TAKE_PHOTO = 1
        private const val REQUEST_WRITE_PERMISSION = 786
        private const val MY_CAMERA_REQUEST_CODE = 100
    }
}